module/Activity/src/Activity/Controller/AdminCategoryController.php

```diff
@@ -0,0 +1,125 @@
+        $categories = $categoryService->getAllCategories();
+
+        return ['categories' => $categories];
+    }
+
+    /**
+     * Add Category.
+     */
+    public function addAction()
+    {
+        $categoryService = $this->getServiceLocator()->get('activity_service_category');
+        $form = $categoryService->getCategoryForm();
+        $request = $this->getRequest();
+        $translator = $this->getServiceLocator()->get('translator');
+
+        if ($request->isPost()) {
+            $form->setData($request->getPost());
+
+            if ($form->isValid()) {
+                $category = $categoryService->createCategory($form->getData());
+
+                if ($category instanceof \Activity\Model\ActivityCategory) {
+                    $message = $translator->translate('The activity category was successfully created!');
+
+                    return $this->redirectWithNotice(true, $message);
+                }
+            }
+        }
+
+        return ['form' => $form, 'action' => $translator->translate('Create Activity Category')];
+    }
+
+    /**
+    * Delete Category.
+    */
+    public function deleteAction()
+    {
+        $request = $this->getRequest();
+        if ($request->isPost()) {
+            $categoryId = (int) $this->params('id');
+            $categoryService = $this->getServiceLocator()->get('activity_service_category');
+
+            $category = $categoryService->getCategoryById($categoryId);
+            if (is_null($category)) {
+                return $this->notFoundAction();
+            }
+
+            $categoryService->deleteCategory($category);
+            return $this->redirect()->toRoute('activity_admin_categories');
+        }
+
+
+        return $this->notFoundAction();
+    }
+
+    /**
+     * Edit Category.
+     */
+    public function editAction()
+    {
+        $categoryId = (int) $this->params('id');
+        $categoryService = $this->getServiceLocator()->get('activity_service_category');
+
+        $category = $categoryService->getCategoryById($categoryId);
+        if (is_null($category)) {
+            return $this->notFoundAction();
+        }
+
+        $categoryService = $this->getServiceLocator()->get('activity_service_category');
+        $form = $categoryService->getCategoryForm();
+        $request = $this->getRequest();
+        $translator = $this->getServiceLocator()->get('translator');
+
+        if ($request->isPost()) {
+            $form->setData($request->getPost());
+
+            if ($form->isValid()) {
+                $category = $categoryService->updateCategory($category, $form->getData());
+
+                if ($category instanceof \Activity\Model\ActivityCategory) {
+                    $message = $translator->translate('The activity category was successfully updated!');
+
+                    return $this->redirectWithNotice(true, $message);
+                }
+            }
```

Comment:
Same here, why not handle the form in the category service?

Reply:
Putting this in the service would also allow you to structure it a bit nicer without nested if statements.

Reply:
Thanks for the feedback! I have no idea why I started doing it in this way, because - as you mentioned - everywhere else we do not use this particular "method".

Reply:
@kokx I have changed these two occurrences, however, I think some of the older parts of the Activity module also need to be changed to reflect this.

In the meantime I also found out why I started using this specific structure. In the Zend 2 documentation the above is used for all form (validation) examples. It is most probable that I unconsciously started doing the same thing.